### PR TITLE
implement sending of 0.5-RTT data

### PR DESCRIPTION
**Pull request by marten-seemann**

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -171,6 +171,16 @@ type Session interface {
 	ConnectionState() tls.ConnectionState
 }
 
+// An EarlySession is a session that is handshaking.
```
Comment (lucas-clemente) [inline interface.go]: Document or at least link to the security guarantees here.

File: server.go [hunk 2]
```diff
@@ -86,19 +88,41 @@ type server struct {
...
 
 	logger utils.Logger
 }
 
-var _ Listener = &server{}
-var _ unknownPacketHandler = &server{}
+var _ Listener = &baseServer{}
+var _ unknownPacketHandler = &baseServer{}
+
+type earlyServer struct{ *baseServer }
```
Comment (lucas-clemente) [inline server.go]: Maybe my Go is rusty, but can't this just be an alias?

File: server.go [hunk 3]
```diff
@@ -126,7 +150,17 @@ func Listen(conn net.PacketConn, tlsConf *tls.Config, config *Config) (Listener,
...
 }
 
-func listen(conn net.PacketConn, tlsConf *tls.Config, config *Config) (*server, error) {
+// ListenEarly works like Listen, but it returns sessions before the handshake completes.
+func ListenEarly(conn net.PacketConn, tlsConf *tls.Config, config *Config) (EarlyListener, error) {
+	s, err := listen(conn, tlsConf, config)
+	if err != nil {
+		return nil, err
+	}
+	s.acceptEarlySessions = true
```
Comment (lucas-clemente) [inline server.go]: Is there a race condition here if a client connects before this is set?

Comment (marten-seemann) [reply server.go]: Right, this should better be a parameter to `listen`.

File: server.go (hunk 2 above)
Comment (marten-seemann) [reply server.go]: I think I need a struct here to define the `Accept()` method, don't I?

**Pull request merged**